src/document_formatting_edit_provider/dbtDocumentFormattingEditProvider.ts

```diff
@@ -53,42 +60,56 @@ export class DbtDocumentFormattingEditProvider
     }
     return [];
   }
-
+ 
   private processDiffOutput(
     document: TextDocument,
     diffOutput: string
   ): TextEdit[] {
     const textEdits: TextEdit[] = [];
     const diffs = parseDiff(diffOutput);
+    const lineCount = document.lineCount;
     diffs.forEach((diff) => {
       diff.chunks.forEach((chunk) => {
+        const lastLineChunk = chunk.newLines
         chunk.changes.forEach((change) => {
           if (this.isAddChange(change)) {
+            console.log(change)
+            // Ensure lines addded are not out of bounds
+            var targetLine = Math.min(change.ln, lineCount);
             textEdits.push(
               TextEdit.insert(
-                document.lineAt(change.ln - 1).range.start,
+                document.lineAt(targetLine - 1).range.start,
                 change.content.slice(1) + "\n"
               )
             );
           }
           if (this.isNormalChange(change)) {
             // Reflect "replace" edits as delete & insert
             // First, delete line
+            console.log(change)
+            
             textEdits.push(
               TextEdit.delete(
                 document.lineAt(change.ln1 - 1).rangeIncludingLineBreak
               )
             );
 
             // Add line
+            // Ensure lines addded are not out of bounds
+            let targetLine = Math.min(change.ln2, lineCount);
+            // Prevent interference at end of chunk
+            if (targetLine == lastLineChunk) {
+              targetLine--
+            }
             textEdits.push(
               TextEdit.insert(
-                document.lineAt(change.ln2 - 1).range.start,
+                document.lineAt(targetLine - 1).range.start,
```

Comment:
Thanks for the review. You're right, let me work out a solution,

In the meantime:
Maybe another more pragmatic solution direction would be just to use the `sqlfmt` without the `--diff` argument,  this would eliminate any buggy post-processing and directly return the formatted file. 
What were the arguments in favor of using the `--diff` approach? Are they still valid when considering this approach currently introduces formatting bugs?

Reply:
This is the api that vscode formatters should support. 

Maybe the fix is to keep track of the lines deleted & inserted and add the increment to the targetLine accordingly as the diff doesn't take into account the already applied changes.

Reply:
Clear, I found a solution for the bugs using the formatter API.
I changed the logic, it now uses the end of a chunk as a maximum bound.
This way you don't have interference between chunks, while also solving the initial bug.

Because the TextEdits are collected in an array and then pushed in one go to the file, the lineCount of the original file is leading. The correct location in the original file to insert a line can be derived by using a maximum of
```const oldBoundChunk = chunk.oldLines + chunk.oldStart - 1;```

`oldLines` are the amount of lines in the chunk, before formatting
`oldStart` is the location of the first line of the chunk in the original file

Using this maximum I set the insertion location as follows: 

```document.lineAt(Math.min(change.ln, oldBoundChunk) - 1).range.start```

I tested against multiple SQL inputs, and the parsing matches the original output from `sqlfmt`

For example: 
```
with
    cte_1 as (
        select 1 as col_1, 1 as col_2, 1 as col_3 union all select 2 as id_1, 2 as id_2, 2 as col_3
    ),  -- cte_1

    cte_2 as (
        select 3 as col_1, 3 as col_2, 3 as col_3
        union all
        select 4 as id_1, 4 as id_2, 4 as col_3
    ),  -- cte_2
    cte_3 as (
        select 5 as col_1, 5 as col_2, 5 as col_3
        union all
        select 6 as id_1, 6 as id_2, 6 as col_3
    ),  -- cte_3

select *
from cte_1
```

Where `--diff` returns multiple chunks (and some change locations are outside of the file bounds)
```
--- source_query
+++ formatted_query
@@ -1,6 +1,8 @@
 with
     cte_1 as (
-        select 1 as col_1, 1 as col_2, 1 as col_3 union all select 2 as id_1, 2 as id_2, 2 as col_3
+        select 1 as col_1, 1 as col_2, 1 as col_3
+        union all
+        select 2 as id_1, 2 as id_2, 2 as col_3
     ),  -- cte_1
 
     cte_2 as (
@@ -15,4 +17,4 @@
     ),  -- cte_3
 
 select *
-from cte_1
\ No newline at end of file
+from cte_1
```

I processed the rest of the remarks; removed log pollution and used inline logic for the `targetLine` instead.